lam4-backend/src/Lam4/Expr/Printer.hs

```diff
@@ -0,0 +1,305 @@
+      _                    -> t ++ s
+    where
+      (spc, rest) = span isSpace s
+
+  closingOrPunctuation :: String -> Bool
+  closingOrPunctuation [c] = c `elem` closerOrPunct
+  closingOrPunctuation _   = False
+
+  closerOrPunct :: String
+  closerOrPunct = ")],;"
+
+parenth :: Doc -> Doc
+parenth ss = doc (showChar '(') . ss . doc (showChar ')')
+
+concatS :: [ShowS] -> ShowS
+concatS = foldr (.) id
+
+concatD :: [Doc] -> Doc
+concatD = foldr (.) id
+
+replicateS :: Int -> ShowS -> ShowS
+replicateS n f = concatS (replicate n f)
+
+-- | The printer class does the job.
+
+class Print a where
+  prt :: Int -> a -> Doc
+
+instance {-# OVERLAPPABLE #-} Print a => Print [a] where
+  prt i = concatD . map (prt i)
+
+instance Print Char where
+  prt _ c = doc (showChar '\'' . mkEsc '\'' c . showChar '\'')
+
+instance Print String where
+  prt _ = printString
+
+instance Print T.Text where
+  prt i x = prPrec i 0 (concatD [doc (showString (T.unpack x))])
+
+printString :: String -> Doc
+printString s = doc (showChar '"' . concatS (map (mkEsc '"') s) . showChar '"')
+
+mkEsc :: Char -> Char -> ShowS
+mkEsc q = \case
+  s | s == q -> showChar '\\' . showChar s
+  '\\' -> showString "\\\\"
+  '\n' -> showString "\\n"
+  '\t' -> showString "\\t"
+  s -> showChar s
+
+prPrec :: Int -> Int -> Doc -> Doc
+prPrec i j = if j < i then parenth else id
+
+instance Print Int where
+  prt _ x = doc (shows x)
+
+instance Print Double where
+  prt _ x = doc (shows x)
+
+instance Print [Expr] where
+  prt _ [] = concatD []
+  prt _ [x] = concatD [prt 0 x]
+  prt _ (x:xs) = concatD [prt 0 x, doc (showString ","), prt 0 xs]
+
+instance Print [Name] where
+  prt _ [] = concatD []
+  prt _ [x] = concatD [prt 0 x]
+  prt _ (x:xs) = concatD [prt 0 x, doc (showString ","), prt 0 xs]
+
+instance Print [Statement] where
+  prt _ [] = concatD []
+  prt _ [x] = concatD [prt 0 x]
+  prt _ (x:xs) = concatD [prt 0 x, doc (showString "\n\n"), prt 0 xs]
+
+instance Print [Decl] where
+  prt _ [] = concatD []
+  prt _ [x] = concatD [prt 0 x]
+  prt _ (x:xs) = concatD [prt 0 x, doc (showString "\n\n"), prt 0 xs]
+
+instance Print [RowTypeDecl] where
+  prt _ [] = concatD []
+  prt _ [x] = concatD [prt 0 x]
+  prt _ (x:xs) = concatD [prt 0 x, doc (showString "\n"), prt 0 xs]
+
+-- type Row a = [(Name, a)] where
+instance Print (Row Expr) where
+  prt _ [] = concatD []
+  prt _ [x] = concatD [prt 0 x]
+  prt _ (x:xs) = concatD [prt 0 x, doc (showString ","), prt 0 xs]
+
+-- Building block for Row
+instance Print (Name, Expr) where
+  prt i (name, expr) = prPrec i 0 (concatD [prt 0 name, doc (showString "="), prt 0 expr])
+
+instance Print Decl where
+  prt i = \case
+    NonRec name expr -> prPrec i 0 (concatD [prt 0 name, prt 0 expr])
+
+    -- Name is actually part of the Predicate.
+    -- doing this quick hack to move it to where it belongs.
+    Rec name (Predicate vars expr ref) -> prPrec i 0 (concatD [prt 0 (Predicate (name:vars) expr ref)])
```

Comment:
Couple of things here (and let's talk more about this after the meeting today if time permits):

1. Re `Decl`: I just remembered that Simala also has a Render / Pretty instance that might be helpful: https://github.com/smucclaw/simala/blob/746a6e0c98923374e55d7590fe30d93b1c9a3bef/src/Simala/Expr/Render.hs#L33 
I would adopt a strategy like the one they take, at least for now, since the intent behind my design of the Decl vs Expr is pretty similar to theirs. In particular, the name in the Decl is not in some sense unique to Predicates. A Decl is basically just a Binding, except that it'd typically be a binding for a declaration (and in the case of Lam4, if I recall correctly, a *top-level* declaration, though this might change in the future).
Sorry, I should have linked you to their Render code earlier.

2. The Predicate example, though, does remind me that I probably have to augment the concrete syntax in the Haskell backend with more type signature / annotation constructs (and similarly with constructs like the FunDecl). The concrete syntax is a bit too 'abstract' in that sense (my original intent had just been to get something that would be easy to desugar even more and evaluate; I hadn't been aiming for bidirectionality).

Reply:
1. I see, I wasn't aware of the Simala AST. That makes total sense, and as I said, I can get the things rendered correctly by just passing some arguments to lower levels. I will in that case just make that more intentional.
2. Yes, I also noticed that. This example for instance
```
// placeholder function for illustration
Integer => Integer
sumIntegers(x, y) = x + y

GIVEN ()
DECIDE `some predicate`
IF sumIntegers(1, 2) > 2
```

becomes this after parsing and prettyprinting:

```
sumIntegers (x, y) = x + y

GIVEN ()
DECIDE `some predicate`
IF sumIntegers (1, 2) > 2
```

Reply:
> 1. I see, I wasn't aware of the Simala AST. That makes total sense, and as I said, I can get the things rendered correctly by just passing some arguments to lower levels. I will in that case just make that more intentional.
> 2. Yes, I also noticed that. This example for instance
> 
> ```
> // placeholder function for illustration
> Integer => Integer
> sumIntegers(x, y) = x + y
> 
> GIVEN ()
> DECIDE `some predicate`
> IF sumIntegers(1, 2) > 2
> ```
> 
> becomes this after parsing and prettyprinting:
> 
> ```
> sumIntegers (x, y) = x + y
> 
> GIVEN ()
> DECIDE `some predicate`
> IF sumIntegers (1, 2) > 2
> ```

I could also write a transformation to a 'more concrete' version of the concrete syntax. I just need to think more about this / play around with this a bit

Reply:
I'll look into this! Probably the most important annotations / metadata to preserve would be the descriptions, if I'm understanding this example correctly. (The Integer and Boolean type annotations in this example should already be preserved.)

Reply:
I've just added a commit to PR #69 -- https://github.com/smucclaw/lam4/pull/69/commits/93e9ec78f7a43b6b1df303dec49bbe8c96259963 --- that partially addresses this (and tweaks the convention re metadata in the Langium grammar).

The new record decl metadata convention (this has also changed for function and predicate decls):
```lam4
/- description: "Info about a lottery." -/
STRUCTURE Lottery {
  -- how much can be won from the jackpot.
  total_jackpot: Integer
  -- whether buying tickets from this lottery is tax deductible.  
  `tax deductible status`: Boolean
}
```
The backend parsing for the `/- description: "Info about a lottery." -/` doesn't work yet --- I want to refactor that part of the Langium grammar first. But the inline descriptions do work:

```
[ TypeDecl
    "Lottery"
    _1
    (RecordDecl
       [ MkRowTypeDecl
           { name = "total_jackpot" _3
           , typeAnnot = BuiltinType BuiltinTypeInteger
           , metadata = Just "-- how much can be won from the jackpot."
           }
       , MkRowTypeDecl
           { name = "`tax deductible status`" _4
           , typeAnnot = BuiltinType BuiltinTypeBoolean
           , metadata =
               Just
                 "-- whether buying tickets from this lottery is tax deductible.  "
           }
       ]
       []
       RecordDeclMetadata
         { transparency = Transparent , description = Nothing })
]
```

Reply:
Makes sense! I just changed the printer based on the new ConcreteSyntax.hs in add-eval branch. As of [51241679](https://github.com/smucclaw/lam4/pull/67/commits/51241679ccfc23202dac4ec8018889247210294f) this 

```haskell
TypeDecl
  (MkName "Lottery" 1)
  (RecordDecl
     [ ( (MkName "total_jackpot" 2) , BuiltinType BuiltinTypeInteger )
     , ( (MkName "`tax deductible status`" 4) , BuiltinType BuiltinTypeBoolean ) ]
     []
     (RecordDeclMetadata Transparent (Just "game where you lose money")))
```

is prettyprinted into this:

```python
/-
  game where you lose money
-/
STRUCTURE Lottery
{
  total_jackpot : Integer
  `tax deductible status` : Boolean
}
```

Reply:
Ah there is one little complication: the `/- ... -/` is no longer permitted for the fields. I.e., it should be 

```
/-
  description:  game where you lose money
-/
STRUCTURE Lottery
{
   -- how much can be won from the jackpot
  total_jackpot : Integer
  
    -- whether buying tickets from this lottery is tax deductible
  `tax deductible status` : Boolean
}
```
But if this new convention doesn't seem like a good idea, please do let me know!